scripts/generateDeepExports.mjs

```diff
@@ -0,0 +1,362 @@
+ * @param {string} targetPackagePath
+ * @param {string} utilityName
+ * @returns {Promise<boolean>}
+ */
+async function hasExtraExports(targetPackagePath, utilityName) {
+  const utilityIndexPath = path.join(targetPackagePath, 'src', utilityName, 'index.ts');
+
+  if (!existsSync(utilityIndexPath)) {
+    return false;
+  }
+
+  const content = await fs.readFile(utilityIndexPath, 'utf8');
+  const contentWithoutBlock = removeExistingBlock(content);
+
+  // Check if there's any meaningful content left (not just empty lines or comments)
+  const meaningfulLines = contentWithoutBlock.split('\n').filter((line) => {
+    const trimmed = line.trim();
+    return trimmed && !trimmed.startsWith('//') && !trimmed.startsWith('/*');
+  });
+
+  return meaningfulLines.length > 0;
+}
+
+/**
+ * Determine the best source package to export from for a utility
+ * @param {string} targetPackagePath
+ * @param {string} utilityName
+ * @param {string[]} sourcePackageChain - Array of source packages in order (e.g., ['x-charts', 'x-charts-pro'])
+ * @returns {Promise<string>} - The package name to export from
+ */
+async function getBestSourcePackage(targetPackagePath, utilityName, sourcePackageChain) {
+  const targetPackageName = path.basename(targetPackagePath);
+
+  const sourcePackageChecks = sourcePackageChain
+    .filter((sourcePackage) => sourcePackage !== targetPackageName) // Skip the current target package
+    .map(async (sourcePackage) => {
+      const sourcePackagePath = path.join(process.cwd(), 'packages', sourcePackage);
+      const hasExtras = await hasExtraExports(sourcePackagePath, utilityName);
+      return hasExtras ? sourcePackage : null;
+    });
+
+  const results = await Promise.all(sourcePackageChecks);
+
+  // Find the first non-null result (prioritizing later packages in the chain)
+  for (let i = results.length - 1; i >= 0; i -= 1) {
+    const result = results[i];
+    if (result !== null) {
+      return result;
+    }
+  }
+
+  // If no source package has extra exports, use the base package
+  return sourcePackageChain[0];
+}
+
+/**
+ * Add re-export block at the beginning of the file
+ * @param {string} content
+ * @param {string} sourcePackageName
+ * @param {string} utilityName
+ * @returns {string}
+ */
+function addReExportBlock(content, sourcePackageName, utilityName) {
+  const blockContent = [
+    BLOCK_IDENTIFIER_LINE_START,
+    `export * from '@mui/${sourcePackageName}/${utilityName}';`,
+    BLOCK_IDENTIFIER_LINE_END,
+    '',
+  ].join('\n');
+
+  return blockContent + content;
+}
+
+/**
+ * Get component directories from src folder
+ * @param {string} packagePath
+ * @param {string[]} ignoreList
+ * @returns {Promise<string[]>}
+ */
+async function getComponentDirectories(packagePath, ignoreList) {
+  const srcPath = path.join(packagePath, 'src');
+
+  if (!existsSync(srcPath)) {
+    return [];
+  }
+
+  const dirents = await fs.readdir(srcPath, { withFileTypes: true });
```

Comment:
Would it be better to `readdir` and handle the `ENOENT ` error if it doesn't exist?

It's one less call. For a lot of files it might be faster. 

Same applies on line 189, and other lines across this script. 

Not a blocker, though. Feel free to do it as a follow-up if you prefer.

Reply:
Plus it's vulnerable to race conditions. Another program may remove the dir in between the `exists` check and the `readdir` call. So you technically have to handle that error anyway. At which point the `exists` check becomes unnecessary. I believe this would be one of the reasons why `exists` was deprecated in the first place.

Reply:
> It's one less call. For a lot of files it might be faster.

The script is quite fast, I don't think there is any concern in that aspect. I've refactored to use catch blocks instead on existsSync however.

Reply:
nitpick, but those catch blocks could potentially hide other errors than just the missing dir.